File: a2/java/src/Layton.java
import java.awt.Point;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import static java.lang.Math.abs;

public class Layton implements Searchable, Comparable<Searchable> {

    protected int[][] board;
    protected List<List<Point>> shapes;
    protected List<Point> current;
    protected List<Point> goal;
    protected double cost;
    private int hash;
    private int heur;

    public Layton(int[][] board, List<List<Point>> shapes, List<Point> current, List<Point> goal) {
        this.board = board;
        this.shapes = shapes;
        this.current = current;
        this.goal = goal;
    }

    @Override
    public boolean isDone() {
        return current.equals(goal);
    }

    @Override
    public int getDistance(Searchable target) {
        return 1;
    }

    @Override
    public double getHeuristic() {
        if (heur == 0) {
            Point p = current.get(0), q = goal.get(0);
            heur = abs(p.x - q.x) + abs(p.y - q.y);
            // Micro-optimization: (3, 5) is the black triangle's bottom-left corner.
            if (p.x < 3 && p.y < 5) {
                heur += (5 - p.y);
            }
        }
        return heur;
    }

    @Override
    public List<Searchable> getMoves() {
        List<Searchable> moves = new ArrayList<>();
        for (int i = 0; i < shapes.size(); i++) {
            for (Point dir : Search.STEPS) {
                Layton m = getMove(i, dir);
                if (m != null) {
                    moves.add(m);
                }
            }
        }
        return moves;
    }

    @Override
    public void setCost(double cost) {
        this.cost = cost;
    }

    @Override
    public double getCost() {
        return cost;
    }

    @Override
    public boolean equals(Object obj) {
        int[][] b = ((Layton) obj).board;
        return Arrays.deepEquals(board, b);
    }

    @Override
    public int hashCode() {
        if (hash == 0) {
            hash = Arrays.deepHashCode(board);
        }
        return hash;
    }

    @Override
    public int compareTo(Searchable o) {
        return Double.compare(getCost(), o.getCost());
    }

    public void print() {
        for (int[] row : board) {
            System.out.println(Arrays.toString(row));
        }
    }

    private Layton getMove(int ind, Point dir) {
        List<Point> shape = shapes.get(ind);
        Point p = new Point();
        for (Point coor : shape) {
            p.setLocation(coor.x + dir.x, coor.y + dir.y);
            if (board[p.y][p.x] != -1 && !shape.contains(p)) {
                return null;
            }
        }
        int[][] nboard = new int[board.length][];
        for (int i = 0; i < board.length; i++) {
            nboard[i] = Arrays.copyOf(board[i], board[i].length);
        }
        List<List<Point>> nshapes = new ArrayList<>(shapes);
        List<Point> nshape = new ArrayList<>();
        Set<Point> moved = new HashSet<>();
        for (Point coor : shape) {
            p = new Point(coor.x + dir.x, coor.y + dir.y);
            if (!moved.contains(coor)) {
                nboard[coor.y][coor.x] = -1;
            }
            nboard[p.y][p.x] = board[coor.y][coor.x];
            nshape.add(p);
            moved.add(p);
        }
        nshapes.set(ind, nshape);
        return new Layton(nboard, nshapes, shape.equals(current) ? nshape : current, goal);
    }

    private static Layton getStart() {
        int[][] board = {
                {-2, -2, -2, -2, -2, -2, -2, -2, -2, -2},
                {-2, -2, -2, -1, -1, 10, 10, -2, -2, -2},
                {-2, -2, -1, -1, -1, 10, 9, -1, -2, -2},
                {-2, 0, 0, -1, -2, 9, 9, -1, -1, -2},
                {-2, 0, 0, -2, -2, 6, 7, -1, -1, -2},
                {-2, 1, 2, 2, 6, 6, 7, 7, 8, -2},
                {-2, 1, 1, 2, -1, 6, 7, 8, 8, -2},
                {-2, -2, -1, 5, 3, 3, 4, 4, -2, -2},
                {-2, -2, -2, 5, 5, 3, 4, -2, -2, -2},
                {-2, -2, -2, -2, -2, -2, -2, -2, -2, -2}
        };
        final List<Point> start = Arrays.asList(new Point(2, 3), new Point(1, 3), new Point(2, 4), new Point(1, 4));
        final List<Point> goal = Arrays.asList(new Point(6, 1), new Point(5, 1), new Point(6, 2), new Point(5, 2));
        List<List<Point>> shapes = new ArrayList<List<Point>>(){{
            add(start);
            add(Arrays.asList(new Point(6, 1), new Point(5, 1), new Point(5, 2)));
            add(Arrays.asList(new Point(6, 2), new Point(5, 3), new Point(6, 3)));
            add(Arrays.asList(new Point(1, 5), new Point(1, 6), new Point(2, 6)));
            add(Arrays.asList(new Point(3, 5), new Point(2, 5), new Point(3, 6)));
            add(Arrays.asList(new Point(5, 4), new Point(5, 5), new Point(5, 6), new Point(4, 5)));
            add(Arrays.asList(new Point(6, 4), new Point(6, 5), new Point(6, 6), new Point(7, 5)));
            add(Arrays.asList(new Point(8, 5), new Point(8, 6), new Point(7, 6)));
            add(Arrays.asList(new Point(3, 7), new Point(3, 8), new Point(4, 8)));
            add(Arrays.asList(new Point(5, 7), new Point(4, 7), new Point(5, 8)));
            add(Arrays.asList(new Point(7, 7), new Point(6, 7), new Point(6, 8)));
        }};
        return new Layton(board, shapes, start, goal);
    }

    public static void main(String[] args) {
        List<String> params = Arrays.asList(args);
        int bfs2 = 3615884;
        int astar2 = 2841892;
        // Run only if explicitly specified.
        if (params.contains("run")) {
            boolean debug = params.contains("debug");
            Searchable start = getStart();
            Search search = new Search(debug);
            // BFS takes 3x longer than astar.
            Search.Result resBfs = search.bfs(start);
            Search.Result resAstar = search.astar(start);
            bfs2 = resBfs.iterations;
            astar2 = resAstar.iterations;
            if (debug) {
                ((Layton) resAstar.goal).print();
                System.out.println("bfs2-cost=" + resBfs.cost);
                System.out.println("astar2-cost=" + resAstar.cost);
            }
        }
        System.out.println("bfs2=" + bfs2);
        System.out.println("astar2=" + astar2);
    }

}

File: a2/layton.py
import numpy as np
from search import (bfs, astar)


class Layton(object):

    def __init__(self, board=None, shapes=None, current=None, goal=None):
        self.board = board
        self.shapes = shapes
        self.current = current
        self.goal = goal
        self._hash = None
        self._h = None

    def __eq__(self, o):
        return np.array_equal(self.board, o.board)

    def __hash__(self):
        self._hash = self._hash or hash(self.board.tostring())
        return self._hash

    def __str__(self):
        return str(self.board)

    def distance(self, target):
        return 1

    def done(self):
        return self.current == self.goal

    def heuristic(self):
        if self._h is None:
            p, q = self.current[0], self.goal[0]
            self._h = abs(p[0] - q[0]) + abs(p[1] - q[1])
        return self._h

    def moves(self):
        for ind, shape in enumerate(self.shapes):
            up = self._move_up(ind)
            if up: yield up
            down = self._move_down(ind)
            if down: yield down
            left = self._move_left(ind)
            if left: yield left
            right = self._move_right(ind)
            if right: yield right

    def _move_up(self, shape_ind):
        return self._move(shape_ind, lambda c: (c[0]+1, c[1]))

    def _move_down(self, shape_ind):
        return self._move(shape_ind, lambda c: (c[0]-1, c[1]))

    def _move_left(self, shape_ind):
        return self._move(shape_ind, lambda c: (c[0], c[1]-1))

    def _move_right(self, shape_ind):
        return self._move(shape_ind, lambda c: (c[0], c[1]+1))

    def _move(self, shape_ind, step_handler, d=False):
        shape = self.shapes[shape_ind]
        for coor in shape:
            step = step_handler(coor)
            if self.board[step] != -1 and step not in shape:
                return
        nboard = np.array(self.board, copy=True)
        nshapes = list(self.shapes)
        nshape = []
        moved = set()
        for coor in shape:
            step = step_handler(coor)
            if coor not in moved:
                nboard[coor] = -1
            nboard[step] = self.board[coor]
            nshape.append(step)
            moved.add(step)
        nshapes[shape_ind] = nshape
        ncurrent = nshape if shape == self.current else self.current
        return Layton(nboard, nshapes, ncurrent, self.goal)


def gen_start():
    board = np.array([
        np.full(10, -2),
        [-2, -2, -2, -1, -1, 10, 10, -2, -2, -2],
        [-2, -2, -1, -1, -1, 10, 9, -1, -2, -2],
        [-2, 0, 0, -1, -2, 9, 9, -1, -1, -2],
        [-2, 0, 0, -2, -2, 6, 7, -1, -1, -2],
        [-2, 1, 2, 2, 6, 6, 7, 7, 8, -2],
        [-2, 1, 1, 2, -1, 6, 7, 8, 8, -2],
        [-2, -2, -1, 5, 3, 3, 4, 4, -2, -2],
        [-2, -2, -2, 5, 5, 3, 4, -2, -2, -2],
        np.full(10, -2),
    ])
    shapes = [
        [(1, 5), (1, 6), (2, 5)],
        [(2, 6), (3, 5), (3, 6)],
        [(3, 1), (3, 2), (4, 1), (4, 2)],
        [(5, 1), (6, 1), (6, 2)],
        [(5, 2), (5, 3), (6, 3)],
        [(5, 4), (4, 5), (5, 5), (6, 5)],
        [(4, 6), (5, 6), (6, 6), (5, 7)],
        [(5, 8), (6, 8), (6, 7)],
        [(7, 3), (8, 3), (8, 4)],
        [(7, 4), (7, 5), (8, 5)],
        [(7, 6), (7, 7), (8, 6)],
    ]
    return Layton(board=board,
                  shapes=shapes,
                  current=[(3, 1), (3, 2), (4, 1), (4, 2)],
                  goal=[(1, 5), (1, 6), (2, 5), (2, 6)])


if __name__ == '__main__':
    start = gen_start()
    # Each of these take 20-30 mins.
    # Use the Java version which the the exact port
    # but finishes in <3 minutes.
    goal, cost, iters = bfs(start, debug=True)
    print('bfs2-cost: %d' % cost)
    print('bfs2-pops: %d' % iters)

    goal, cost, iters = astar(start, debug=True)
    print('astar2-cost: %d' % cost)
    print('astar2-pops: %d' % iters)

